Add `keep_in_fp32_modules` support

younesbelkada opened the pull request:
# What does this PR do?

This PR partially addresses #20287 - although half-precision and int8 conversion work extremely well for most of the models, for some architectures (e.g. T5) the casting leads in a drastic performance degradation. 

This can be fixed by manually force-casting some modules in `float32`. For FLAN-T5, @larsmennen and @navjotts have found out that keeping only these weights in `fp32` enables to run largest models in fp16 or int8 with no performance degradation.

This PR introduces a new utils in `from_pretrained` method, termed as `keep_in_fp32_modules` that partially addresses this issue.

How this util works? For T5:
```
from transformers import T5ForConditionalGeneration

model = T5ForConditionalGeneration.from_pretrained("t5-small", torch_dtype=torch.float16, keep_in_fp32_modules=["wo"])
print(model.decoder.block[0].layer[2].DenseReluDense.wo.weight.dtype)
>>> torch.float32
```

When using `keep_in_fp32_modules` , `low_cpu_mem_usage` needs to be force-set to `True`. This is because if `low_cpu_mem_usage=False`, it is the function from pytorch `_load_from_state_dict` that is called under the hood on each sub-module. This function calls `copy_` from Pytorch which seems to keep the tensor in its native `dtype` regardless the `dtype` of the input module

```
import torch

param = torch.Tensor([0.1, 0.2, 0.3]).to(torch.float16)
to_copy_param = torch.Tensor([0.2, 0.1, 0.3]).to(torch.float32)

param.copy_(to_copy_param)
print(param.dtype)
>>> torch.float16
```

Keeping this as a draft for now as this util needs to be manually patched with fixes such as https://github.com/huggingface/transformers/issues/20287#issuecomment-1342219429 , otherwise users will encounter issues about incompatible `dtype` between input and weights

cc @sgugger

younesbelkada committed:
add `keep_in_fp32_modules` support

younesbelkada commented:
What about adding hooks on each converted module, that will take care of converting the input / output to the correct `dtype` ?

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1968,6 +1977,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         load_in_8bit = kwargs.pop("load_in_8bit", False)
         load_in_8bit_threshold = kwargs.pop("load_in_8bit_threshold", 6.0)
         load_in_8bit_skip_modules = kwargs.pop("load_in_8bit_skip_modules", None)
+        keep_in_fp32_modules = kwargs.pop("keep_in_fp32_modules", None)
```
This should be a class attribute on the model class, not yet another kwarg in `from_pretrained`.

younesbelkada replied:
Agreed and fixed in 2bd00b3 - but I think we still need a kwarg here, imagine if a user does not want to upcast these modules?

younesbelkada replied:
Or we can just enable it if `torch_dtype != torch.float32`

sgugger replied:
They can just adapt the class variable in that case, but they shouldn't not do this, that's the whole point.

younesbelkada replied:
I see, I am convinced now!

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1982,6 +1992,13 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             elif not low_cpu_mem_usage:
                 raise ValueError("Passing along a `device_map` requires `low_cpu_mem_usage=True`")
 
+        if keep_in_fp32_modules is not None and not low_cpu_mem_usage:
+            # Force `low_cpu_mem_usage` to be set to `True` - check the PR:
+            logger.warning(
+                "The argument `keep_in_fp32_modules` is used, force-enabling `low_cpu_mem_usage` to load the model"
+            )
+            low_cpu_mem_usage = True
```
Shouldn't be force-set here.

younesbelkada replied:
proposed something in [2bd00b3](https://github.com/huggingface/transformers/pull/20683/commits/2bd00b3eaf1316cecbc76b9ef281b9d8f8a3e26d)

sgugger reviewed (commented):
Some initial comments:

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20683). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
pass it as class attribute

younesbelkada committed:
few modifs

- make tests `slow`
- fix logic

younesbelkada committed:
better logic

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2280,7 +2288,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             else:
                 loaded_state_dict_keys = [k for k in state_dict.keys()]
-            if low_cpu_mem_usage:
+            if low_cpu_mem_usage or cls._keep_in_fp32_modules is not None:
```
I don't really like this solution, it forces the user to have `accelerate` installed if they want to use t5, even for a small model. Note that the original issue comes only for large models. So it's fine to not upcast those weights in fp32 for small models. Proposed an alternative in [a62a594](https://github.com/huggingface/transformers/pull/20683/commits/a62a594e6a97ce32ed5b070ff894dfe2b54af0af)

younesbelkada committed:
fix failing test

younesbelkada committed:
`bfloat16` support

younesbelkada commented:
As suggested in #20287 / model loaded in bfloat16 should keep their weights in `bfloat16` and not cast them in `fp32`. This is addressed in  e3498da

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -611,7 +612,14 @@ def _load_state_dict_into_meta_model(
+            if (
+                keep_in_fp32_modules is not None
+                and any(module_to_keep_in_fp32 in param_name for module_to_keep_in_fp32 in keep_in_fp32_modules)
+                and dtype != torch.bfloat16
```
```suggestion
                and dtype == torch.float16
```
This is specifically for float16 dtypes (I don't think we get the int8 here?)

younesbelkada replied:
Yes, totally!

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2276,11 +2286,25 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+            if cls._keep_in_fp32_modules is not None:
+                if not is_accelerate_available() or torch_dtype == torch.bfloat16:
+                    use_keep_in_fp32_modules = False
+                    logger.warning(
+                        " `_keep_in_fp32_modules` is not set to `None` and you don't have `accelerate` installed",
+                        " it is recommended to have `accelerate` installed in this case `pip install accelerate`.",
+                    )
```
The warning should only be trigerred when `torch_dtype == torch.float16`.

younesbelkada replied:
Should be fixed in [8014c34](https://github.com/huggingface/transformers/pull/20683/commits/8014c34e7aa8e1ed3475d4a002e3be46eb3e258e)

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2299,6 +2323,10 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         with ContextManagers(init_contexts):
             model = cls(config, *model_args, **model_kwargs)
 
+        keep_in_fp32_modules = model._keep_in_fp32_modules
+        if keep_in_fp32_modules is not None:
+            low_cpu_mem_usage = True
```
Shouldn't this use `use_keep_in_fp32_modules` here? Also should go before so the test at line 2307 can be simplified.

younesbelkada replied:
I simplified the tests in 966cc06 but I think that we still need `keep_in_fp32_modules = model._keep_in_fp32_modules` as it is used later on in line 2342

younesbelkada replied:
I think you are right: 0f75387

younesbelkada replied:
Actually it seems that it's more tricky than that, putting it on top will result in some failing tests should be fixed in 243e6b5

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2534,11 +2568,25 @@ def _fix_key(key):
+                # upcast in fp32 if any
+                target_dtype = dtype
+                if keep_in_fp32_modules is not None and any(
+                    module_to_keep_in_fp32 in key for module_to_keep_in_fp32 in keep_in_fp32_modules
```
should also add a test of `dtype` being `float16` here.

younesbelkada replied:
added in 8014c34

younesbelkada committed:
Update src/transformers/modeling_utils.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
fix

younesbelkada committed:
simplify tests

younesbelkada committed:
simplify tests

younesbelkada committed:
fix test

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2276,11 +2286,25 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                if not is_accelerate_available() and torch_dtype == torch.float16:
+                    use_keep_in_fp32_modules = False
+                    logger.warning(
+                        " `_keep_in_fp32_modules` is not set to `None` and you don't have `accelerate` installed"
+                        " it is recommended to have `accelerate` installed in this case `pip install accelerate`.",
```
The warning is not really useful for a user that will just use `from_pretrained`. I'd say something like:
"For stability purposes, it is recommended to have `accelerate` installed when using this model in `torch.flaot16`, please install it with `pip install accelerate`."

younesbelkada replied:
Thanks for the suggestion! Fixed in 1e80f14

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2299,6 +2323,10 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
+        if use_keep_in_fp32_modules:
+            low_cpu_mem_usage = True
+        keep_in_fp32_modules = model._keep_in_fp32_modules
```
Let's set it to `[]` here if it's not `None`, so that we don't have to check again layer on.

younesbelkada replied:
This should be addressed in [cb89c42](https://github.com/huggingface/transformers/pull/20683/commits/cb89c42fd2cd05aacb56c15462869c229faaf3fa)

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2534,11 +2568,27 @@ def _fix_key(key):
+                        set_module_8bit_tensor_to_device(
+                            model, key, "cpu", torch.empty(*param.size(), dtype=target_dtype)
+                        )
+        elif keep_in_fp32_modules is not None and state_dict is not None:
+            for key in state_dict:
+                if any(module_to_keep_in_fp32 in key for module_to_keep_in_fp32 in keep_in_fp32_modules):
+                    state_dict[key] = state_dict[key].to(torch.float32)
```
This is not useful as with `torch.load_state_dict`, the weights are converted to the dtype inside the model. So it's the model dtype that you should fix here.

Also this removes the necessity for an Accelerate warning above, no?

younesbelkada replied:
Yes! Should be addressed in [cb89c42](https://github.com/huggingface/transformers/pull/20683/commits/cb89c42fd2cd05aacb56c15462869c229faaf3fa)

younesbelkada committed:
modify message

younesbelkada committed:
more checks

larsmennen commented:
Thanks @younesbelkada and @sgugger !! Tested this locally; can confirm this works with patch 1&2 from https://github.com/huggingface/transformers/issues/20287#issuecomment-1342219429

The only problem I encountered is that in: 

https://github.com/huggingface/transformers/blob/0f75387d633619b8c0cf6955b07dac54d2e17473/src/transformers/modeling_utils.py#L2326

You get an error as `keep_in_fp32_modules` is an unexpected keyword to the underlying model class (locally i just added it quickly to test). Do you want to add this in so people can use it in their model class to determine where to apply patches like 1&2? Or alternatively don't pass it on and then people can just query the dtype.

younesbelkada committed:
fix failing tests

younesbelkada commented:
Thanks so much @larsmennen for confirming that the tests pass! We should be close merging this 💪 
I am not sure but I think that your failing test should be fixed with my latest commit ( cb89c42 ) but I am not sure, could you try again with the latest commit? 🙏

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2070,6 +2070,10 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         loading_info = None
 
+        # Keep in fp32 modules
+        keep_in_fp32_modules = None
```
@larsmennen the keyword has been added here if this is what you meant

younesbelkada committed:
add more conditions

- add `is_accelerate_available`
- fixes pipleine tests that failed

larsmennen commented:
hmm that doesn't fix it. I think you just need to pop the argument from model_kwargs, otherwise it gets passed to the underlying model (i'm assuming you don't want that? but cmiiw)

I.e. after

https://github.com/huggingface/transformers/blob/7d47df2e52c6b55d8f19e5b2bd8b5e472a4f0a82/src/transformers/modeling_utils.py#L1981

if you add

```py
keep_in_fp32_modules = kwargs.pop("keep_in_fp32_modules", None)
```

I tested w/ that modification on top of [7d47df2](https://github.com/huggingface/transformers/pull/20683/commits/7d47df2e52c6b55d8f19e5b2bd8b5e472a4f0a82) and that works! Thanks for the quick action @younesbelkada ! 🙏

younesbelkada commented:
@larsmennen how are you loading your model ? The description above is slightly misleading as initially the plan was to add a kwarg when loading the model as follows:
```
from transformers import T5ForConditionalGeneration

model = T5ForConditionalGeneration.from_pretrained("t5-small", torch_dtype=torch.float16, keep_in_fp32_modules=["wo"])
```
but now this is not needed, you should just load your model like:
```
from transformers import T5ForConditionalGeneration

model = T5ForConditionalGeneration.from_pretrained("t5-small", device_map="auto", load_in_8bit=True])
```

larsmennen commented:
@younesbelkada ah i see! I was passing the kwarg yes, so that explains.

sgugger reviewed (approved):
Almost there, just one last comment and we should be good to merge! Thanks!

younesbelkada committed:
add suggestions

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2276,11 +2290,14 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 dtype_orig = cls._set_default_torch_dtype(torch_dtype)
 
+            # Check if `_keep_in_fp32_modules` is not None
+            use_keep_in_fp32_modules = cls._keep_in_fp32_modules is not None and is_accelerate_available()
```
Agreed! Proposed your suggestions in 1d21843

younesbelkada commented:
@larsmennen this PR will be merged as soon as all the tests will be green ! 
Would you mind opening a PR addressing your suggestions (patch 1 & 2 from the discussion at #20287 )?

younesbelkada committed:
Update src/transformers/modeling_utils.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
fix failing `bnb` test

younesbelkada committed:
Merge branch 'add-fp32-modules' of https://github.com/younesbelkada/transformers into add-fp32-modules

younesbelkada committed:
add last safety checker

younesbelkada commented:
All slow tests from T5 (and BLOOM just in case we didn't break anything else) pass 🟢 
Merging once the CI tests are green

Oxi84 commented:
I tried this one, latest version of transformers (27.4), cuda 10.2 and I get this error:

`model1a_CPU = T5ForConditionalGeneration.from_pretrained(model_path, low_cpu_mem_usage=True,torch_dtype=torch.float16, keep_in_fp32_modules=["wo"]).to("cuda")               
TypeError: __init__() got an unexpected keyword argument 'keep_in_fp32_modules'`

sgugger commented:
`keep_in_fp32_modules` is not an argument you can pass to `from_pretrained`, this is all done internally.

younesbelkada commented:
You need to do somthing like:

```python
from transformers import T5ForConditionalGeneration

T5ForConditionalGeneration._keep_in_fp32_modules = ["wo"]

# your code here
```

sgugger commented:
Except this is already done for T5 ;-)
